subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c

```diff
@@ -208,11 +226,235 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+	}
+
+	/* Ensure path is null-teminated */
+	path[sizeof(path) - 1] = '\0';
+
+	/* Retrieve file size */
+	rc = fs_mgmt_impl_filelen(path, &file_len);
+
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* Encode the response. */
+	err = 0;
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "len");
+	err |= cbor_encode_uint(&ctxt->encoder, file_len);
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "rc");
+	err |= cbor_encode_int(&ctxt->encoder, MGMT_ERR_EOK);
+
+	if (err != 0) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return 0;
+}
+#endif
+
+#if defined(CONFIG_FS_MGMT_CHECKSUM_HASH_SUPPORTED)
+/**
+ * Command handler: fs hash/checksum (read)
+ */
+static int
+fs_mgmt_file_hash_checksum(struct mgmt_ctxt *ctxt)
+{
+	char path[CONFIG_FS_MGMT_PATH_SIZE + 1];
+	char type[HASH_CHECKSUM_TYPE_SIZE + 1];
+	char *output = NULL;
+	unsigned long long len = ULLONG_MAX;
+	unsigned long long off = 0;
+	CborError err;
+	size_t file_len;
+	int rc;
+	struct fs_file_t file;
+	const struct hash_checksum_mgmt_group *group = NULL;
+
+	const struct cbor_attr_t size_attr[] = {
+		{
+			.attribute = "type",
+			.type = CborAttrTextStringType,
+			.addr.string = type,
+			.len = sizeof(type),
+			.nodefault = true
+		},
+		{
+			.attribute = "name",
+			.type = CborAttrTextStringType,
+			.addr.string = path,
+			.len = sizeof(path),
+		},
+		{
+			.attribute = "off",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &off,
+			.nodefault = true
+		},
+		{
+			.attribute = "len",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &len,
+			.nodefault = true
+		},
+		{ 0 },
+	};
+
+	rc = cbor_read_object(&ctxt->it, size_attr);
+
+	if (rc != 0 || len == 0) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Search for supported hash/checksum */
+	group = hash_checksum_mgmt_find_handler(type);
+
+	if (group == NULL) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Check provided offset is valid for target file */
+	rc = fs_mgmt_impl_filelen(path, &file_len);
+
+	if (rc != 0) {
+		return MGMT_ERR_ENOENT;
+	}
+
+	if (file_len <= off) {
+		/* Requested offset is larger than target file size */
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Allocate a buffer for the output data */
+	output = k_malloc(group->output_size);
+
+	if (output == NULL) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	/* Open file for reading and pass to hash/checksum generation function */
+	fs_file_t_init(&file);
+	rc = fs_open(&file, path, FS_O_READ);
+
+	if (rc != 0) {
+		k_free(output);
+		return MGMT_ERR_ENOENT;
+	}
+
+	/* Seek to file's desired offset, if parameter was provided */
+	if (off != 0) {
+		rc = fs_seek(&file, off, FS_SEEK_SET);
+
+		if (rc != 0) {
+			fs_close(&file);
+			k_free(output);
+			return MGMT_ERR_EINVAL;
+		}
+	}
+
+	/* Calculate hash/checksum using function */
+	file_len = 0;
+	rc = group->function(ctxt, &file, output, &file_len, len);
+
+	fs_close(&file);
+
+	/* Encode the response */
+	err = cbor_encode_text_stringz(&ctxt->encoder, "rc");
+	err |= cbor_encode_int(&ctxt->encoder, rc);
```

Comment:
No, don't. We should stop using "rc" for reporting non SMP errors and success does not OK to be sent back.

Reply:
Done